### PR TITLE
Added content rules

### PR DESCRIPTION
**Pull request by mloginov**
https://bugtracker.codiodev.com/issue/codio-15374/Linter-Improvement-Add-checks

### CODE REVIEW

File: src/helpers.js [hunk 1]
```diff
@@ -62,3 +64,60 @@ export const openEditor = async () => {
...
+  } catch {
+    return false
+  }
+}
+
+export const checkExternalUrlExists = async (url) => {
+  try {
+    const command = `wget -q --method=HEAD ${url}`
+    const res = await window.codioIDE.remoteCommand.run(command)
+    return res.code === 0
```
Comment (destitutus) [inline src/helpers.js:81]: I am not sure if `wget` is installed in all boxes
The common command not found code is `127` - maybe we could add extra logic here

File: src/helpers.js [hunk 2]
```diff
@@ -62,3 +64,60 @@ export const openEditor = async () => {
...
+  } catch {
+    return false
+  }
+}
+
+export const promiseAllSeries = (arr) => {
```
Comment (destitutus) [inline src/helpers.js:87]: what is the difference between `Promise.all` and the function?

File: src/rules/page/content/resource-not-found.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import {
+  checkResourceUrlExists,
+  checkExternalUrlExists,
+  collectAllLinks
+} from '../../../helpers'
+
+export default {
+  // {section, content, assessmentIds, assessmentsById}
+  action: async ({content}) => {
+    const promises = collectAllLinks(content).filter(linkInfo => !!linkInfo.url)
```
Comment (destitutus) [inline src/rules/page/content/resource-not-found.js:11]: I have seen that `collectAllLinks`  could return `null` items - it is possible place for `Cannot read properties of null (reading 'url')`

File: src/helpers.js (hunk 2 above)
Comment (mloginov) [reply src/helpers.js:87]: Promise.all runs all request together. promiseAllSeries - one by one

File: src/rules/page/content/resource-not-found.js (hunk 3 above)
Comment (mloginov) [reply src/rules/page/content/resource-not-found.js:11]: not null, but 

`{url: null}`

File: src/rules/page/content/img-alt.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+export default {
+  // {section, content, assessmentIds, assessmentsById}
+  action: ({content}) => {
+    let ruleFailed = false
+    const walkTokens = (token) => {
+      if (token.type === 'image') {
+        const fileNameAlt = token.href
+          .replace(/^.*[\\/]/, '') // get filename
+          .replace(/\.[^/.]+$/, '') // cut extension
```
[the remark this answers is not in the document]
Comment (mloginov) [reply src/rules/page/content/img-alt.js:11]: no

File: src/helpers.js (hunk 1 above)
Comment (mloginov) [reply src/helpers.js:88]: added logs to console if fetch or remote command failed or not succeeded

File: src/helpers.js [hunk 5]
```diff
@@ -62,3 +64,68 @@ export const openEditor = async () => {
...
+    console.log(e.message)
+    return false
+  }
+}
+
+export const checkExternalUrlExists = async (url) => {
+  try {
+    const command = `wget -q --method=HEAD ${url}`
+    const res = await window.codioIDE.remoteCommand.run(command)
+    if (res.status !== 200) {
```
Comment (destitutus) [inline src/helpers.js:85]: does `run` return `status` field with http code?

Comment (mloginov) [reply src/helpers.js:85]: yes

**Review by destitutus — APPROVED** [review]

**Pull request merged**